### PR TITLE
feat(twitter): add an NFT tab in profile page

### LINKED ISSUE #3751: [Demand] Enhance tab list on the profile page
## Catalogue

- [ ] UI/UX
- [x] Mask Core Feature
- [ ] Mask Plugin Feature
- [ ] Security/Cryptography Feature
- [ ] Other
  - /_List it here_/

## Part 1: Introduction and Background

image

### LINKED ISSUE #3767: [Demand] Add twitter identifier resolve service
## Catalogue

- [ ] UI/UX
- [ ] Mask Core Feature
- [x] Mask Plugin Feature (Collectible)
- [ ] Security/Cryptography Feature
- [ ] Other
  - /_List it here_/

## Part 1: Introduction and Background

+ ENS
+ 3Box

### LINKED ISSUE #3808: [Demand] Add an NFTs tab on profile page
## Catalogue

- [ ] UI/UX
- [ ] Mask Core Feature
- [x] Mask Plugin Feature
- [ ] Security/Cryptography Feature
- [ ] Other
  - /_List it here_/

## Part 1: Introduction and Background

Add an NFTs tab on the profile page if it's possible to resolve the Twitter account info to a wallet account.

+ [ ] Add an NFTs tab on the profile page (@lelenei). 
+ [ ] #3877
    + If the Twitter nickname is a valid ENS domain name, use it to resolve an address.
    + If it's possible to find an ENS domain name from the bio, then use it to resolve to an address.
    + If there is a valid ethereum address in the bio text, use it as the address.
    + Use `LowerCase(Twitter id) + '.eth'` to resolve to an address.
    + If the self-profile page is viewing, it fallback to the address of the currently selected wallet.
+ [ ] Fetch collectible list with the resolved address (@UncleBill).

<b>This feature is not related to the currently selected network type.</b>

### PR DESCRIPTION
**Pull request by lelenei**
closes #3751
closes #3767 
closes #3808

### CODE REVIEW

**Comment by septs** [comment]
plz rename this pr title, thanks

**Comment by UncleBill** [comment]
Improve layout of `<CollectibleList/>`

image

image

File: packages/maskbook/src/social-network/types.ts
Comment (Jack-Works) [inline packages/maskbook/src/social-network/types.ts:125]:
```suggestion
            /** Inject UI to the Profile page */
```

File: packages/maskbook/src/social-network-adaptor/twitter.com/injection/EnhancedProfile.tsx
Comment (guanbinrui) [inline packages/maskbook/src/social-network-adaptor/twitter.com/injection/EnhancedProfile.tsx:161]:
```suggestion
          <div key="nfts" className={classes.tab}>
              <div className={classNames(classes.button, active ? classes.hot : '')} onClick={onClick}>
                  NFTs
                  {active ? <div className={classes.active} /> : null}
              </div>
          </div>
```

File: packages/maskbook/src/social-network-adaptor/twitter.com/utils/user.ts
Comment (guanbinrui) [inline packages/maskbook/src/social-network-adaptor/twitter.com/utils/user.ts:18]:
```suggestion
export const getTwitterNickname = () => {
```

File: packages/maskbook/src/social-network-adaptor/twitter.com/injection/EnhancedProfile.tsx
Comment (guanbinrui) [inline packages/maskbook/src/social-network-adaptor/twitter.com/injection/EnhancedProfile.tsx:180]: A single file for this hook.

File: packages/web3-shared/src/hooks/useResolveEns.ts
Comment (guanbinrui) [inline packages/web3-shared/src/hooks/useResolveEns.ts:10]:
```suggestion
    }, [web3, name])
```

Comment (UncleBill) [reply packages/web3-shared/src/hooks/useResolveEns.ts]: Nice catch.

**Comment by UncleBill** [comment]
@DimensionDev/qa It's ready for test.

**Comment by Jack-Works** [comment]
> @DimensionDev/qa It's ready for test.

Please wait for my review before merge thanks!!

File: packages/maskbook/src/social-network-adaptor/twitter.com/injection/EnhancedProfile.tsx
Comment (Jack-Works) [inline packages/maskbook/src/social-network-adaptor/twitter.com/injection/EnhancedProfile.tsx:136]: Please cleanup addEventListener added in the `bind` by `useEffect`

File: packages/maskbook/src/social-network-adaptor/twitter.com/injection/EnhancedProfile.tsx
Comment (Jack-Works) [inline packages/maskbook/src/social-network-adaptor/twitter.com/injection/EnhancedProfile.tsx:183]: Please wrap it in `useEffect`

**Review by Jack-Works — CHANGES_REQUESTED** [review]

File: packages/maskbook/src/social-network-adaptor/twitter.com/injection/EnhancedProfile.tsx
Comment (Jack-Works) [inline packages/maskbook/src/social-network-adaptor/twitter.com/injection/EnhancedProfile.tsx:188]: You didn't return the cleanup function

File: packages/maskbook/src/social-network-adaptor/twitter.com/injection/EnhancedProfile.tsx
Comment (Jack-Works) [inline packages/maskbook/src/social-network-adaptor/twitter.com/injection/EnhancedProfile.tsx:123]: Run `useEffect` inside a loop wrong

File: packages/maskbook/src/social-network-adaptor/twitter.com/injection/EnhancedProfile.tsx
Comment (Jack-Works) [inline packages/maskbook/src/social-network-adaptor/twitter.com/injection/EnhancedProfile.tsx:112]: This function must starts with `use` since it use `useEffect` inside.

**Review by Jack-Works — CHANGES_REQUESTED** [review]

File: packages/maskbook/src/social-network-adaptor/twitter.com/injection/EnhancedProfile.tsx
Comment (septs) [inline packages/maskbook/src/social-network-adaptor/twitter.com/injection/EnhancedProfile.tsx:45]: plz extract the type definition as interface

File: packages/maskbook/src/social-network-adaptor/twitter.com/injection/EnhancedProfile.tsx
Comment (septs) [inline packages/maskbook/src/social-network-adaptor/twitter.com/injection/EnhancedProfile.tsx]:
```suggestion
        const onClick = () => {
```

File: packages/maskbook/src/social-network-adaptor/twitter.com/injection/EnhancedProfile.tsx
Comment (Jack-Works) [inline packages/maskbook/src/social-network-adaptor/twitter.com/injection/EnhancedProfile.tsx:204]:
```suggestion
    useEffect(() => {
        return MaskMessage.events.profileNFTsPageUpdate.on((data) => {
            setShow(data.show)
        })
    }, [])
```

**Review by Jack-Works — CHANGES_REQUESTED** [review]

File: packages/maskbook/src/social-network-adaptor/twitter.com/injection/EnhancedProfile.tsx
Comment (Jack-Works) [inline packages/maskbook/src/social-network-adaptor/twitter.com/injection/EnhancedProfile.tsx:211]:
```suggestion
        return MaskMessage.events.profileNFTsPageUpdate.on((data) => {
```

File: packages/maskbook/src/extension/options-page/DashboardComponents/CollectibleList/index.tsx
Comment (Jack-Works) [inline packages/maskbook/src/extension/options-page/DashboardComponents/CollectibleList/index.tsx:168]: Remove this part if you don't have your own styles. Use only `withClasses` is enough.

File: packages/maskbook/src/extension/options-page/DashboardComponents/CollectibleList/index.tsx
Comment (Jack-Works) [inline packages/maskbook/src/extension/options-page/DashboardComponents/CollectibleList/index.tsx:178]:
```suggestion
    const classes = props.classes ?? {}
```

File: packages/maskbook/src/social-network-adaptor/twitter.com/injection/EnhancedProfile.tsx
Comment (Jack-Works) [inline packages/maskbook/src/social-network-adaptor/twitter.com/injection/EnhancedProfile.tsx:127]:
```suggestion
function useReplaceState(handler: () => void) {
```

**Review by Jack-Works — CHANGES_REQUESTED** [review]

**Review by Jack-Works — APPROVED** [review]

**Comment by septs** [comment]
plz resolve confliect

**Review by Jack-Works — APPROVED** [review]

**Review by Jack-Works — DISMISSED** [review]

**Review by Jack-Works — DISMISSED** [review]

File: packages/maskbook/src/components/InjectedComponents/EnhancedProfile.tsx
Comment (guanbinrui) [inline packages/maskbook/src/components/InjectedComponents/EnhancedProfile.tsx:121]: Move this to `packages/maskbook/src/utils/hooks`.

File: packages/maskbook/src/components/InjectedComponents/EnhancedProfile.tsx
Comment (guanbinrui) [inline packages/maskbook/src/components/InjectedComponents/EnhancedProfile.tsx:129]:
```suggestion
    useEffect(() => {
        window.addEventListener('locationchange', handler)
        return () => window.removeEventListener('locationchange', handler)
    }, [handler])
```

File: packages/maskbook/src/components/InjectedComponents/EnhancedProfile.tsx
Comment (guanbinrui) [inline packages/maskbook/src/components/InjectedComponents/EnhancedProfile.tsx:192]: Should filter dialog URLs out. It's better to implement these logics in a  `<EnhancedProfileAtTwttier />` component.

Learn more: packages/maskbook/src/social-network-adaptor/twitter.com/collecting/getSearchedKeyword.ts

File: packages/maskbook/src/components/InjectedComponents/EnhancedProfile.tsx
Comment (guanbinrui) [inline packages/maskbook/src/components/InjectedComponents/EnhancedProfile.tsx:46]:
```suggestion
    useEffect(() => {
        return MaskMessage.events.profileNFTsTabUpdated.on((data) => {
            onUpdated()
        })
    }, [])
  
    useLocationChange(() => {
        MaskMessage.events.profileNFTsTabUpdated.sendToLocal('reset')
    })
```

File: packages/maskbook/src/components/InjectedComponents/EnhancedProfileTab.tsx
Comment (guanbinrui) [inline packages/maskbook/src/components/InjectedComponents/EnhancedProfileTab.tsx:33]:
```suggestion
    children?: ReactElement
```

File: packages/maskbook/src/social-network-adaptor/twitter.com/utils/selector.ts
Comment (Jack-Works) [inline packages/maskbook/src/social-network-adaptor/twitter.com/utils/selector.ts:20]: You should not rely on `aria-label` because it is different in different natural languages, for example, it will not work for Chinese or Japanese users because they have `aria-label="プロファイルタイムライン"`.

This suggestion also applies to other new selectors in this PR.

**Review by Jack-Works — CHANGES_REQUESTED** [review]

**Review by Jack-Works — DISMISSED** [review]

File: packages/maskbook/src/components/InjectedComponents/EnhancedProfile.tsx
Comment (guanbinrui) [inline packages/maskbook/src/components/InjectedComponents/EnhancedProfile.tsx:36]:
```suggestion
    useLocationChange(() => {
        MaskMessage.events.profileNFTsTabUpdated.sendToLocal('reset')
    })
```

**Review by Jack-Works — APPROVED** [review]

**Review by guanbinrui — DISMISSED** [review]

File: packages/maskbook/src/social-network-adaptor/twitter.com/injection/EnhancedProfileTab.tsx
Comment (Jack-Works) [inline packages/maskbook/src/social-network-adaptor/twitter.com/injection/EnhancedProfileTab.tsx:60]: Why not `100%`?

**Review by guanbinrui — APPROVED** [review]

**Pull request merged**
